src/utils/simplifier.js

```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
   switch (value.kind) {
     case "!":
       return negate(realm, value.args[0]);
+    case "||":
+    case "&&":
+      let x = simplify(realm, value.args[0]);
+      let y = simplify(realm, value.args[1]);
+      if (x instanceof AbstractValue && x.equals(y)) return x;
+      return AbstractValue.createFromLogicalOp(realm, (value.kind: any), x, y, value.expressionLocation);
+    case "==":
+    case "!=":
+    case "===":
+    case "!==":
+      return simplifyEquality(realm, value);
     default:
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (x instanceof ConcreteValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : xy) op undefined/null" to just "cond" or "!cond"
+    let [cond, xx, xy] = x.args;
+    invariant(cond instanceof AbstractValue); // otherwise the the conditional should not have been created
+    if (op === "===" || op === "!==") {
+      // if xx === undefined && xy !== undefined then cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightNotBeUndefined() && !xy.mightBeUndefined())
+        return op === "===" ? cond : negate(realm, cond);
```

Comment:
Change requested: The original equality was guaranteed to yield a Boolean (true or false). This possibly replaces it with just `cond`. Is `cond` really guaranteed to be a Boolean? 

Prepacking the following doesn't seem to force a "Booleanization" of a numeric condition.

```js
  let x = __abstract("number", 42);
  y = x ? 1 : 2;
```

Similarly, I'd hope that `negate` always yields a Boolean in this setting, but I am not sure. Would be nice to see invariants for all this, especially around simplification rules.

Reply:
Great catch. Will fix.